enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,264 @@
+
+## Summary
+
+When working with bare-metal OpenShift clusters, due to the absence of storage provisioners like those in the cloud, there is a lot of manual work to be done for consuming local storage from nodes.
+The idea is to automate the manual steps that are required for consuming local storage in an OpenShift cluster by extending the Local Storage Operator (LSO).
+
+## Motivation
+
+To facilitate the consumption of locally-attached storage devices on OpenShift nodes in a cloud-native way.
+
+### Goals
+
+- Automatic detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Should respond to the attach/detach events of the disks/devices.
+- Should have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enables automated discovery andn provisioning of local volume based PVs.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+    // will be looking for during detection
+    DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+    Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    MinDeviceCount int `json:"minDeviceCount"`
```

Comment:
This is actually per node AND per storageclass. So technically user can create more local PVs by specifying a different storageclass. So this value is not really on per-node basis. cc @jsafrane

Reply:
What is the behaviour when a node does not have requested `minDeviceCount`? Is LSO supposed to not provision any PVs at all on the node or is it supposed to provision just number of available devices and log an error event?

Reply:
yes no pvs will be provisioned if the findings didn't match the `minDeviceCount`.

Reply:
Then please add it to the comment

Reply:
@leseb can you reconfirm that this is the behaviour we decided on? Just for my peace of mind :grimacing:

Reply:
If `minDeviceCount` is not provided, we should probably go for all devices matching the rule. And yes, if `minDeviceCount` doesn't strictly match on all nodes, nothing will be done.